DUI3-DX/Converters/ArcGIS/Speckle.Converters.ArcGIS3/ArcGISConversionContextStack.cs

```diff
@@ -1,18 +1,99 @@
 using System.Diagnostics.CodeAnalysis;
-using ArcGIS.Core.Geometry;
+using ArcGIS.Core.Data.DDL;
+using ArcGIS.Core.Data;
+using ArcGIS.Desktop.Core;
+using ArcGIS.Desktop.Framework.Threading.Tasks;
 using ArcGIS.Desktop.Mapping;
 using Speckle.Converters.Common;
 
 namespace Speckle.Converters.ArcGIS3;
 
+public record ArcGISDocument(Project Project, Map Map, Uri SpeckleDatabasePath);
+
 // POC: Suppressed naming warning for now, but we should evaluate if we should follow this or disable it.
 [SuppressMessage(
   "Naming",
   "CA1711:Identifiers should not have incorrect suffix",
   Justification = "Name ends in Stack but it is in fact a Stack, just not inheriting from `System.Collections.Stack`"
 )]
-public class ArcGISConversionContextStack : ConversionContextStack<Map, Unit>
+public class ArcGISConversionContextStack : ConversionContextStack<ArcGISDocument, ACG.Unit>
 {
-  public ArcGISConversionContextStack(IHostToSpeckleUnitConverter<Unit> unitConverter)
-    : base(MapView.Active.Map, MapView.Active.Map.SpatialReference.Unit, unitConverter) { }
+  public ArcGISConversionContextStack(IHostToSpeckleUnitConverter<ACG.Unit> unitConverter)
+    : base(
+      new ArcGISDocument(Project.Current, MapView.Active.Map, EnsureOrAddSpeckleDatabase()),
+      MapView.Active.Map.SpatialReference.Unit,
+      unitConverter
+    ) { }
+
+  private const string FGDB_NAME = "Speckle.gdb";
+
+  public static Uri EnsureOrAddSpeckleDatabase()
+  {
+    return AddDatabaseToProject(GetDatabasePath());
+  }
+
+  public static Uri GetDatabasePath()
+  {
+    try
+    {
+      var parentDirectory = Directory.GetParent(Project.Current.URI);
+      if (parentDirectory == null)
+      {
+        throw new ArgumentException($"Project directory {Project.Current.URI} not found");
+      }
+      var fGdbPath = new Uri(parentDirectory.FullName);
+      return new Uri($"{fGdbPath}/{FGDB_NAME}");
+    }
+    catch (Exception ex)
+      when (ex
+          is IOException
+            or UnauthorizedAccessException
+            or ArgumentException
+            or NotSupportedException
+            or System.Security.SecurityException
+      )
+    {
+      throw;
+    }
+  }
+
+  public static Uri AddDatabaseToProject(Uri databasePath)
+  {
+    // Create a FileGeodatabaseConnectionPath with the name of the file geodatabase you wish to create
+    FileGeodatabaseConnectionPath fileGeodatabaseConnectionPath = new(databasePath);
+    // Create actual database in the specified Path unless already exists
+    try
+    {
+      Geodatabase geodatabase = SchemaBuilder.CreateGeodatabase(fileGeodatabaseConnectionPath);
```

Comment:
When we need to create it and why dispose it immediately if we won't use in this scope?

Reply:
It creates a database on a hard drive, here we just need to make sure it exists and ArcGIS project is linked to its path